docs/examples/MondrianTreeRegressor.md

```diff
@@ -0,0 +1,273 @@
+    X_train = np.linspace(-np.pi, np.pi, n_points)
+    y_train = np.sin(X_train)
+    X_test = np.linspace(-6.0, 6.0, 100)
+    y_test = np.sin(X_test)
+    return np.expand_dims(X_train, axis=1), y_train, np.expand_dims(X_test, axis=1), y_test
+
+X_train, y_train, X_test, y_test = generate_toy_data()
+plt.plot(X_train.ravel(), y_train, "ro")
+plt.show()
+```
+
+
+![png](../mondrian_tree/plot1.png)
+
+
+## Plotting decision boundaries using ERT's
+
+Let us now scikit-learn's ExtraTreeRegressor to train on the generated toy data, predict on some unseen data and plot decision boundaries in the 1-D space. Also, we set the `max_depth` parameter to 2, which means there can be a maximum of 4 decision boundaries in the 1-D space.
+
+
+```python
+def plot_1D_decision_boundaries(tree_reg, plt):
+    x_l, x_u = plt.xlim()
+    colors = cycle("bgrcmyk")
+    split_thresh = tree_reg.tree_.threshold
+    split_thresh = sorted(split_thresh[split_thresh != -2])
+    split_thresh = [x_l] + split_thresh + [x_u]
+    for s in split_thresh:
+        plt.axvline(s, color="r", linestyle="-")
+
+    for x_l, x_u in zip(split_thresh[1:], split_thresh[:-1]):
+        plt.fill_between((x_l, x_u), -1.5, 1.5, alpha=0.2, color=next(colors))
+
+etr = ExtraTreeRegressor(random_state=2, max_depth=2)
+etr.fit(X_train, y_train)
+y_pred = etr.predict(X_test)
+plt.plot(X_train.ravel(), y_train, "ro")
+plt.plot(X_test.ravel(), y_pred)
+plt.ylim((-1.5, 1.5))
+plot_1D_decision_boundaries(etr, plt)
+plt.show()
+```
+
+![png](../mondrian_tree/plot2.png)
+
+The blue line represents the mean prediction in every region of the decision space.
+
+We notice that as we move away from the training data
+
+1. The predicted mean remains constant.
+2. It is determined by the decision split at the bounds of the space.
+
+Ideally, as move away from the training data we are unsure about the target value to predict and would like to fall back on some prior mean. The mondrian tree solves this problem in a very intelligent way.
+
+##  Mondrian Tree
+
+### Train time
+
+At every node, the split threshold and feature is decided independently of the target or the decrease in impurity! Yes, that is right.
+
+1. The split feature index `f` is drawn with a probability proportional to `u_b[f] - l_b[f]` where `u_b` and `l_b` and the upper and lower bounds of all the features.
+2. After fixing the feature index, the split threshold \(\delta\) is then drawn from a uniform distribution with limits `l_b`, `u_b`.
+
+The intuition being that a feature that has a huge difference between the bounds is likelier to be an "important" feature.
+
+At every node, in a decision tree, the mean and impurity are stored (or class probabilities in case of classification). In addition to this, every node `j` in a mondrian tree tracks two other things.
+
+1. The upper and lower bounds of all the features in that particular node.
+2. The time of split, which is drawn from an exponential with mean \(\sum_{f=1}^D(u_b[f] - l_b[f])\)
```

Comment:
Figure 5.1 highlights difference between mondrian tree and vanilla decision tree.

Another way to think of "Time of split" is the following: imagine that edges between parent and child nodes are associated with a non-negative weight (we need a special parent(root) node). Then the time of split at a node is the sum of the weights along the path from parent(root) to the node. If weights are all 1, the time of split is the depth of the node. So we can think of time of split as "weighted depth".